ADR/ADR-002-PyBind11-For-CPP-Bindings.md

```diff
@@ -0,0 +1,167 @@
+ * call C functions: python code can now call C functions and manipulate C data structures as if they were native Python objects
+
+Using CFFI usally requires to write a certain amount of wrapping code to hide foreign structures and to make the usage of the library more pythonic.
+Exposing complex mechanisms (like an C++ iterator) requires writing and remapping intermediate C structures. 
+The wrapping python code usually has to explicitly care about lifetime of objects. Python itself has no concept of ownership as it is considered in C/C++.
+
+```python
+from cffi import FFI
+ffi = FFI()
+ffi.cdef("int strlen(const char *s);")
+libc = ffi.dlopen("libc.so.6")  # or 'msvcrt.dll' on Windows
+
+length = libc.strlen(ffi.new("char[]", b"hello"))
+print(length)  # Output: 5
+```
+
+
+#### Option 2: PyBind11
+
+pybind11 is a C++ header-only library that enables seamless binding of C++ code to Python. It allows you to expose C++ classes, functions, and data structures to Python with minimal boilerplate.
+The python functionaly is directly expressed in C++ by
+ 
+ * writing a C++ wrapper using pybind11 macros (like PYBIND11_MODULE).
+ * compile the C++ code into a Python extension module (usually done via the buildsystem, e.g. CMake)
+ * import and use the C++ code directly from Python
+
+```cpp
+#include <pybind11/pybind11.h>
+
+int add(int a, int b) {
+    return a + b;
+}
+    
+PYBIND11_MODULE(example, m) {
+    m.def("add", &add, "A function that adds two numbers");
+}
+```
+
+Then to find the module in python, the `PYTHONPATH` can be extended to point to `<builddir>/lib` (or `lib64`):
+
+```python
+import example
+print(example.add(2, 3))  # Output: 5
+```
+
+PyBind11 offers compatibility with most of C++ `std` (iterators, exceptions, smart pointers, variant/tuples, ...) and commonly used modules like NumPy.
+
+#### Option 3: nanobind
+
+Similar to PyBind11 (almost the same interface) but with the purpose of
+  
+  * being faster & having less overhead than pybind11
+  * producing smaller binary sizes
+
+while explicitly having reduced compatibility with `std` and third party modules.
+It is not header-only and requires linking to a shared or static library.
+
+[ ] Listed all seriously considered alternatives 
+[ ] Provided brief descriptions of each option
+[ ] Included pros and cons for each alternative  
+
+### Analysis
+TBD
+
+ Cons Pybind:
+ * well tested and widely used (i.e. TensorFlow, PyTorch)
+
+
+[ ] Compared options against relevant criteria (cost, performance, maintainability, risk, etc)  
+[ ] Explained trade-offs and their implications
+[ ] Referenced any prototypes, benchmarks, or research conducted (if applicable)  
+[ ] Considered both immediate and long-term impacts
+
+## Decision
+
+[ ] Stated the chosen option (clearly and unambiguously)
+[ ] Avoided implementation details (unless crucial to understanding the decision)
+[ ] Included any conditions or limitations on the decision
+
+### Related Decisions
```

Comment:
A related decision (or decision-to-be) is presumably "how do we build python2rust or cpp2rust interfaces" -- there are already cases of each (cascade and earthkit-hydro are python libs with rust usage, and there is a mars client rust cpp thingy based on corrosion ), and we may want to mention it here, especially in the sense like "I have a c++ lib with pybind-based interface, now I would like a rust interface as well, do I start from scratch?". 

I recall vaguely there was some argument "cffi is good because you write the c interface once and then every other language can utilize that plain c interface" -- we may perhaps mention it somewhere here as well and eg refute or explain why we think other arguments outweigh it, etc

Reply:
I think that this discussion went in circles a little bit. That was certainly the motivating factor of CFFI - that you do the C interface once, and can build python, fortran and rust interfaces on top.

And I think that is still a relevant factor here - if you are wanting a Fortran interface you should be going via a C interface, and may as well use CFFI.

But if just wanting Rust and Python, the semantics of Rust are closer to that of C++ than a C interface really permits straightforwardly. And in general, doing both a pybind11 interface and a cxxbridge interface is faster, and less error prone, than writing a proper C interface.